### ISSUE TITLE
[EventListener] Add persistence for continue file

### ISSUE THREAD

**Issue by franckc** [OriginProtocol/origin#852]
A. Currently the event listener uses a local "continue" file on disk to save the last block it processed.
Because we are not using persistent volume, every time we deploy the event-listener, it restarts at the initial "Origin epoch block" and it takes about 30min to 1 hr to catch up to latest block (that will degrade over time as the blockchain gets longer). This implies an indexing delay of up to 1 hr. Soon we want to start relying on the back-end data to power the DApp and this behavior won't be good enough.

B. In addition, we are in the process of integrating the event-listener with the notification server. Upon receiving events of interest, the event-listener calls the notification server webhook which results in the user getting a notification in their browser. We should avoid sending to the user duplicate notifications for the same event, which means:
 - at any point of time, a single event-listener should be active
 - if/when reprocessing events (for example for re-indexing purposes), we should make sure to disable the notification webhook call in the listener.

Here are some options that come to mind to solve both A. and B.:
 1. Store the event-listener state in the DB rather than in a local file
 2. Update the event-listener deployment to use a persistent volume so that the continue file gets preserved across event-listener deploys.
 3. Store the state in zookeeper (or equivalent in GCP)

I'm leaning towards option 1 above. The state stored in the DB table would be used as a "lease". I imagine the table schema could be something like this:
 - listener_type: useful for ability to run multiple listeners in parallel (for example for one-off re-indexing) or in the future if we decide to split the event-listener into multiple instances, each handling separate actions (ex: writing to DB vs calling notifier vs external webhooks, etc...)
 - listener_id: unique id that identifies the instance of the listener holding the lease for the listener_type.
 - block_number: last block processed by listener
 - lease_expiry: date at which the lease expires. If listener does not renew the lease by that date, another listener instance can grabs it.

Any thoughts ?

CC @tomlinton @sparrowDom @DanielVF @micahalcorn

**Comment by DanielVF** [comment]
An option to store the listener block state in a DB/redis makes plenty of sense, since that's one less addtional persistence system we need to deal with.

However! I think that not sending duplicate messages should be the responsibility of the notification server, not the listener. We gain a lot of robustness to external chaos on the listener side by not guaranteeing once-only-delivery. Particularly since we don't have a once-only guarantee from upstream from the listener - we might get a block that doesn't get included. Also, one day someone (not tom) will doof up a config file or infra change and the notification server will get duplicate old events anyway.

Also, being able to rerun old listings through the listener gives us a very convenient way to upgrade our database/search index contents if/when we update the data that origin.js returns.

**Comment by franckc** [comment]
Yes good point that strong guarantees about only-once delivery on the event-listener side will be hard...

Indeed, we could have the notification server making sure it does not send duplicate notifications. That means we'll need a way to keep track what events sent by the event-listener have already been processed in the past by the notifier. Options for that could be storing some sort of unique event-id in Redis or in a DB table. That should be a small amount of data and we are not at scale so I don't think we would need to worry about a scheme to purge this data periodically anytime soon...

In terms of priority I would think:
P1. Persist event-listener state in DB
P2. Prevent sending duplicates in notification server

**Comment by franckc** [comment]
Closing this issue.
- Implemented listener DB persistence in PR #874
- Filed separate issue #875 for notifications server to not send duplicates.
